{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 2B - Capsule vs Capsule Transcriptome:\n",
    "\n",
    "Here we plot the data comparing the two capsule samples - the capsule sample purified from the E. coli sample alone, and the sample purified from the Mixed sample.  The plot generates shows a scatterplot where each point is plotted at the RPM (Reads per Million) value of the sample from the E. coli sample and the Mixed sample. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bk-root\">\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-logo bk-logo-small bk-logo-notebook\"></a>\n",
       "        <span id=\"1002\">Loading BokehJS ...</span>\n",
       "    </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(\"1002\");\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.2.min.js\"];\n",
       "  const css_urls = [];\n",
       "  \n",
       "\n",
       "  const inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    if (force === true) {\n",
       "        display_loaded();\n",
       "      }} else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"1002\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.2.min.js\"];\n  const css_urls = [];\n  \n\n  const inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    if (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"1002\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from bokeh.io import push_notebook, show, output_notebook\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we will read our data frames into a file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Here is the list of unique samples for which there are dataframe files: \n",
      "\n",
      "['Bsub_Lysate', 'Ecoli_Cap', 'Ecoli_Lysate', 'Mix_Cap', 'Mix_Lysate']\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df_dir = '../..//dataframe_files/'\n",
    "prefix = '230117Li_'\n",
    "df_dict = dict()\n",
    "names = []\n",
    "for file in listdir(df_dir):\n",
    "    if file.endswith('_dataframe.txt'):\n",
    "        name = file[file.find(prefix)+len(prefix):]\n",
    "        name = name[:name.find('_1')]\n",
    "        new_df  = pd.read_csv(''.join([df_dir,file]))\n",
    "        new_df['Length'] = new_df['Stop'] - new_df['Start']\n",
    "        df_dict[name] = new_df\n",
    "        names.append(name)\n",
    "print(f'\\nHere is the list of unique samples for which there are dataframe files: \\n\\n{names}\\n')\n",
    "names_dict = {'Ecoli_Cap':'E. coli Capsule', 'Mix_Cap':'Mixture Capsule', 'Ecoli_Lysate':'E. coli Lysate', \n",
    "              'Mix_Lysate':'Mixture Lysate', 'Bsub_Lysate':'B. subtilis Lysate'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some helper functions to help us with the plotting:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def into_RPKM(df_dict, name):\n",
    "    sample_df = df_dict[name].copy()\n",
    "    sample_df = sample_df.rename(columns = {'Counts':name})\n",
    "    sample_df[name] = sample_df[name]/(sum(sample_df[name])/1000000) #RPM\n",
    "    sample_df[name] = sample_df[name].div(sample_df['Length']) #RPKM\n",
    "    \n",
    "    return sample_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Comparative abundance plots:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = 'Ecoli_Cap' # capsule - e coli\n",
    "y = 'Mix_Cap' # capsule - mixture sample\n",
    "\n",
    "x_df = into_RPKM(df_dict, x)\n",
    "y_df = into_RPKM(df_dict, y)\n",
    "plot_df = x_df.merge(y_df)\n",
    "plot_df.drop(plot_df[plot_df['Name'] == 'error'].index, inplace = True)\n",
    "plot_df.drop(plot_df[plot_df[x] == 0].index, inplace = True)\n",
    "plot_df.drop(plot_df[plot_df[y] == 0].index, inplace = True)\n",
    "cap_ecoli_same_df = deepcopy(plot_df)\n",
    "genome_map = {'bsubtilis':'B. subtilis','capsule':'E. coli',\n",
    "            'ecoli':'E. coli'}\n",
    "cap_ecoli_same_df['Genome'] = cap_ecoli_same_df['Genome'].map(genome_map)\n",
    "dot_source = ColumnDataSource(cap_ecoli_same_df)\n",
    "\n",
    "theil_sen_regressor = TheilSenRegressor()\n",
    "theil_sen_regressor.fit(np.log(np.asarray(plot_df[x])).reshape(-1, 1), np.log(np.asarray(plot_df[y])))\n",
    "\n",
    "fit_x_vals = np.sort(np.asarray(plot_df[x]))\n",
    "fit_y_vals = np.exp(theil_sen_regressor.predict(np.log(fit_x_vals).reshape(-1,1)))\n",
    "fit_r2 = r2_score(plot_df[y], np.exp(theil_sen_regressor.predict(np.log(np.asarray(plot_df[x])).reshape(-1,1))))\n",
    "line_source = ColumnDataSource({\"x_line\": fit_x_vals, \"y_line\": fit_y_vals,\n",
    "                                \"Name\": [f'Fit, R^2 = {fit_r2:.3f}' for _ in fit_x_vals]})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets check out what Genomes we have to work with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['B. subtilis', 'E. coli']\n"
     ]
    }
   ],
   "source": [
    "print(list(cap_ecoli_same_df.Genome.unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Noe lets make the plot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "name_1 = names_dict[x].replace('\\n', '')\n",
    "name_2 = names_dict[y].replace('\\n', '')\n",
    "title_str = f\"{name_1} vs. {name_2} RPKM\"\n",
    "font = \"Arial\"\n",
    "hover = HoverTool(tooltips = [('Name', '@Name')])\n",
    "mapper = CategoricalColorMapper(palette=[\"skyblue\",\"sandybrown\"], factors=[\"B. subtilis\", \"E. coli\"])\n",
    "p = figure(title = title_str, y_axis_type = 'log',\n",
    "           x_axis_type = 'log', tools=[hover, 'wheel_zoom', 'box_zoom', 'save', 'reset'],\n",
    "           x_axis_label = f'{names_dict[x]} Sample RPKM',\n",
    "           y_axis_label = f'{names_dict[y]} Sample RPKM',\n",
    "           plot_width = 1200, plot_height = 1200)\n",
    "data = p.circle(x, y, source = dot_source ,fill_alpha=0.35, size=15,\n",
    "         line_alpha = 0.05, color={'field': 'Genome', 'transform': mapper},\n",
    "          legend_field = 'Genome')\n",
    "\n",
    "#fit = p.line(\"x_line\", \"y_line\", source = line_source, line_width=6, line_alpha = 0.5, line_color = \"orange\")\n",
    "p.title.text_font_size = '35pt'\n",
    "p.title.align = \"center\"\n",
    "p.title.text_font = font\n",
    "p.xaxis.axis_label_text_font_size = \"30pt\"\n",
    "p.xaxis.axis_label_text_font = font\n",
    "p.xaxis.axis_label_text_font_style = \"bold\"\n",
    "p.yaxis.axis_label_text_font_size = \"30pt\"\n",
    "p.yaxis.axis_label_text_font = font\n",
    "p.yaxis.axis_label_text_font_style = \"bold\"\n",
    "\n",
    "p.xaxis.major_label_text_font_size = \"25pt\"\n",
    "p.yaxis.major_label_text_font_size = \"25pt\"\n",
    "p.xgrid.grid_line_alpha = 0.5\n",
    "p.ygrid.grid_line_alpha = 0.5\n",
    "\n",
    "#legend = Legend(items=[\n",
    "#    (f'Observed RPKM of a given gene',   [data]),\n",
    "#    (f'Linear Fit R^2: {fit_r2:.3f}', [fit]),\n",
    "#], location=(380, 200), label_text_font_size = \"25px\",\n",
    "#   border_line_alpha = 1)\n",
    "#p.add_layout(legend)\n",
    "p.legend.label_text_font_size = '35pt'\n",
    "p.legend.glyph_height = 50\n",
    "p.legend.glyph_width = 90\n",
    "p.legend.label_height = 50\n",
    "p.legend.label_width = 50\n",
    "p.legend.background_fill_color = 'ghostwhite'\n",
    "p.legend.location = (120, 700)\n",
    "p.legend.border_line_width = 2\n",
    "html = file_html(p, CDN, f\"{x}_vs_{y}\")\n",
    "with open( f\"./{x}_vs_{y}.html\", 'w') as f:\n",
    "    f.write(html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can plot the relative abundances of each genome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "samples = ['Mix_Cap', 'Mix_Lysate', 'Ecoli_Cap', 'Ecoli_Lysate', 'Bsub_Lysate']\n",
    "genome_map = {'bsubtilis':'B. subtilis','capsule':'E. coli', 'ecoli':'E. coli'}\n",
    "\n",
    "real_names = []\n",
    "bar_plot_data = dict()\n",
    "genomes = list(cap_ecoli_same_df.Genome.unique())\n",
    "for g in genomes:\n",
    "    bar_plot_data[g] = []\n",
    "\n",
    "for n in samples:\n",
    "    real_names.append(names_dict[n])\n",
    "    data_df = into_RPKM(df_dict, n)\n",
    "    df_copy = deepcopy(data_df)\n",
    "    df_copy['Genome'] = df_copy['Genome'].map(genome_map)\n",
    "    for g in genomes:\n",
    "        arr = bar_plot_data[g]\n",
    "        arr.append(sum(df_copy[df_copy.Genome == g][n])/sum(df_copy[n]))\n",
    "        bar_plot_data[g] = arr\n",
    "bar_plot_data['samples'] = real_names\n",
    "\n",
    "mapper = CategoricalColorMapper(palette=[\"skyblue\", \"sandybrown\"], factors=[\"B. subtilis\", \"E. coli\"])\n",
    "hover = HoverTool(tooltips = [('Name', '@Name')])\n",
    "p = figure(y_range = [0, 1.25], x_range = real_names, height = 1000, width = 1400, \n",
    "           title=\"Percent Reads Mapping to Each Genome\")\n",
    "p.vbar_stack(genomes, x = 'samples', width=0.9, source = bar_plot_data,\n",
    "             color = [\"skyblue\", \"sandybrown\"],\n",
    "             legend_label = genomes)\n",
    "\n",
    "p.y_range.start = 0\n",
    "p.x_range.range_padding = 0.1\n",
    "p.xgrid.grid_line_color = None\n",
    "p.axis.minor_tick_line_color = None\n",
    "p.outline_line_color = None\n",
    "p.legend.location = (360, 675)\n",
    "p.legend.label_text_font_size = '35pt'\n",
    "p.legend.orientation = \"horizontal\"\n",
    "\n",
    "p.title.text_font_size = \"40pt\"\n",
    "p.title.align = 'center'\n",
    "\n",
    "p.xaxis.major_label_text_font_size = \"35pt\"\n",
    "\n",
    "p.yaxis.axis_label=\"Percentage Mapped\"\n",
    "p.yaxis.axis_label_text_font_size = \"35pt\"\n",
    "p.yaxis.major_label_text_font_size = \"35pt\"\n",
    "p.xaxis.axis_label_text_font_style = \"bold\"\n",
    "p.yaxis.axis_label_text_font_style = \"bold\"\n",
    "\n",
    "p.legend.glyph_height = 50\n",
    "p.legend.glyph_width = 90\n",
    "p.legend.label_height = 50\n",
    "p.legend.label_width = 50\n",
    "\n",
    "html = file_html(p, CDN, f\"Percent Genome Plot {'_'.join(samples)}.html\")\n",
    "with open(f\"Percent Genome Plot {'_'.join(samples)}.html\", 'w') as f:\n",
    "    f.write(html)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also create a 2d plot comparing the relative abundances of each genome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_2d_plot = {\"sample_ids\": ['Mix_Cap', 'Mix_Lysate', 'Ecoli_Cap', 'Ecoli_Lysate', 'Bsub_Lysate'], \"B. subtilis\":[], \"E. coli\":[], \"real_names\": [],\n",
    "             \"colors\": ['lightseagreen', 'lightsalmon', 'lightpink', 'khaki', 'darkturquoise']}\n",
    "genome_map = {'bsubtilis':'B. subtilis','capsule':'E. coli', 'ecoli':'E. coli'}\n",
    "\n",
    "for n in df_2d_plot[\"sample_ids\"]:\n",
    "    df_2d_plot[\"real_names\"].append(names_dict[n])\n",
    "    data_df = into_RPKM(df_dict, n)\n",
    "    df_copy = deepcopy(data_df)\n",
    "    df_copy['Genome'] = df_copy['Genome'].map(genome_map)\n",
    "    for g in list(df_copy.Genome.unique()):\n",
    "        df_2d_plot[g].append((sum(df_copy[df_copy.Genome == g][n])/sum(df_copy[n])))\n",
    "\n",
    "mapper = CategoricalColorMapper(palette=[\"skyblue\", \"sandybrown\"], factors=[\"B. subtilis\", \"E. coli\"])\n",
    "hover = HoverTool(tooltips = [('Sample', '@real_names')])\n",
    "p = figure(title = \"Percent Reads Mapped to Each Genome\", y_axis_type = 'log',\n",
    "           x_axis_type = 'log', tools=[hover, 'wheel_zoom', 'box_zoom', 'save', 'reset'],\n",
    "           x_axis_label = \"Percent Mapping to B. Subtilis Genome\",\n",
    "           y_axis_label = \"Percent Mapping to E. coli Genome\",\n",
    "           plot_width = 1200, plot_height = 1200, y_range = [0.00001, 2.5], x_range = [0.00001, 2.5],)\n",
    "data = p.circle(\"B. subtilis\", \"E. coli\", source = ColumnDataSource(df_2d_plot), size=60, \n",
    "                legend_field = \"real_names\", fill_color = \"colors\", line_color = \"colors\")\n",
    "\n",
    "p.legend.label_text_font_size = '35pt'\n",
    "\n",
    "p.title.text_font_size = \"30pt\"\n",
    "p.title.align = 'center'\n",
    "\n",
    "p.xaxis.axis_label_text_font_size = \"30pt\"\n",
    "p.xaxis.major_label_text_font_size = \"30pt\"\n",
    "\n",
    "p.yaxis.axis_label_text_font_size = \"30pt\"\n",
    "p.yaxis.major_label_text_font_size = \"30pt\"\n",
    "p.xaxis.axis_label_text_font_style = \"normal\"\n",
    "p.yaxis.axis_label_text_font_style = \"normal\"\n",
    "\n",
    "\n",
    "p.legend.glyph_height = 100\n",
    "p.legend.glyph_width = 90\n",
    "p.legend.label_height = 100\n",
    "p.legend.label_width = 100\n",
    "p.legend.location = (200, 200)\n",
    "\n",
    "html = file_html(p, CDN, \"2d_Genome_plot_v2.html\")\n",
    "with open(\"2d_Genome_plot_v2.html\", 'w') as f:\n",
    "    f.write(html)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
